### ISSUE TITLE
Replicating between databases causes SG to go OOM

### ISSUE THREAD

**Issue by b-smets** [couchbase/sync_gateway#2607]
### Sync Gateway version
We are currently running a 1.5 master build: 
http://latestbuilds.hq.couchbase.com/couchbase-sync-gateway/1.5.0/1.5.0-381/
We do this in order to work around https://github.com/couchbase/sync_gateway/issues/2587  
Accessing / results in:
```{"couchdb":"Welcome","vendor":{"name":"Couchbase Sync Gateway","version":1.5},"version":"Couchbase Sync Gateway/1.5.0(381;6a1598b)"}```

### Operating system
Ubuntu 14.04

### Config file
Regular config file with 8 databases and the following memory tweaks:
```
              "feed_type": "TAP",
              "rev_cache_size": 10,
              "cache": {
                "channel_cache_max_length": 10,
                "channel_cache_min_length": 10
	        }
```
Note: We added these only after the first OOMs were encountered.

### Log output
https://gist.github.com/b-smets/50f07bfee28349a4cf3b113eafa7d6e6

In order to work around https://github.com/couchbase/sync_gateway/issues/1648 and shrink our bucket size, we attempted moving all documents of our big bucket into a new bucket, hoping the obsolete attachments would stay behind. The big bucket contains about 1.3M items (according to our Couchbase Web interface) of which are 2507 json documents with 1-5 attachments (<2MB) each, and the rest is obsolete garbage. The target bucket is empty.

We tried running the following command:
```
curl -X POST --header 'Content-Type: application/json' --header 'Accept: application/json' -d '{ "source": "<sourcebucket>",  "target": "<targetbucket>" }' 'http://localhost:4985/_replicate'
```
This command ran for approx. 5 minutes before resulting in replication terminated response from the server. However, the replication was still going strong and items were being transferred between the buckets. After the replication finished, we checked the number of documents in the target bucket but this contained only about 350 documents so it didn't process everything.
Any attempt to run the replication a second time now results in an out of memory error (see gist) after a couple of minutes. No new documents are being transferred.

Other things we tried:
- Use "async": true => Same OOM errors
- Use "continuous": true => This results in nothing happening. No documents being transfered but no oom.
- Replicate into the target database on another SG instance => Same OOM

**Comment by tleyden** [comment]
Thanks for the clear issue report!

One thing that you might try that would help isolate the OOM is to run [sg-replicate](https://github.com/couchbaselabs/sg-replicate) as a standalone tool rather than as being embedded in Sync Gateway.   That would help uncover where the OOM is coming from.

Also, if you can run [sgcollect_info](https://github.com/couchbase/sync_gateway/blob/master/tools/sgcollect_info) and privately share your collected files via email that might also reveal something about where the memory is going.

**Comment by tleyden** [comment]
@b-smets I got the logs you sent via email -- thanks for that. In the `nohup.out` log, I'm seeing this:

```
2017/06/01 13:55:46 [33mReplicate: [0m[db-local] event: &{FETCH_CHANGES_FEED_SUCCEEDED {[{390 d0ce403a ...
```

followed by thousands / tens of thousands of change entries.  There should only be **50** change entries, based on the sample config:

```
{
    "changes_feed_limit": 50,
    "continuous_retry_time_ms": 15000,
    "replications":{
        "db-local":{
            "source_url": "http://localhost:4985",
            "source_db": "db",
            "target_db": "db-copy",
            "lifecycle": "oneshot"
        },
        "grocery-sync-local":{
            "source_url": "http://localhost:4985",
            "target_url": "http://sync.couchbasecloud.com",
            "source_db": "grocery-sync",
            "target_db": "grocery-sync",
            "lifecycle": "continuous",
            "channels": ["lists", "items"]
        }
    }
}
```

what value for `changes_feed_limit` are you using?  (if any).  It looks like if no `changes_feed_limit` is supplied in the config, then it will use a `changes_feed_limit` of **0**, which will request all changes in the entire database in one single request, and be likely to cause the OOM issues you are seeing.

I wasn't able to find out the changes limit from looking at the sync gateway logs you sent, but can you try again with a config that explicitly sets the `changes_feed_limit` to **50** or a smaller number?  

If it's still happening, can you look for log lines in Sync Gateway like:

```
2017-06-02T15:35:13.025-07:00 HTTP:  #001: GET /db/_changes?limit=5  (ADMIN)
2017-06-02T15:35:43.491-07:00 HTTP:  #002: GET /db/_changes?limit=0  (ADMIN)
```

and see what is being passed as a `limit` parameter?

**Comment by tleyden** [comment]
Digging into this some more, it looks like that even with a config with:

```
{
    "changes_feed_limit": 50,
}
```

it's ignoring it and using `changes_feed_limit=0`

**Comment by tleyden** [comment]
@b-smets if you have the ability to [rebuild Sync Gateway](https://github.com/couchbase/sync_gateway/blob/master/docs/BUILD.md) yourself, can you retest the [feature/issue_2615_default_changes_feed_limit](https://github.com/couchbase/sync_gateway/tree/feature/issue_2615_default_changes_feed_limit) branch and see if it improves your situation?

Otherwise a new build should be up to grab in the next few days after the PR is reviewed and merged.

**Comment by tleyden** [comment]
@b-smets I'm going to proactively close this issue and assume it's fixed by https://github.com/couchbase/sync_gateway/issues/2616.  If you retest it after that change and see the same issue, we can re-open this issue.

**Comment by tleyden** [comment]
You can pickup the build from http://latestbuilds.hq.couchbase.com/couchbase-sync-gateway/1.5.0/1.5.0-407/

**Comment by b-smets** [comment]
Glad to hear you were able to find out what could be causing the issue.
I will try to deploy this new SyncGateway on our staging environment this week so I can resume testing and will keep you posted.

**Comment by b-smets** [comment]
@tleyden 
I was able to get this to work fine now with the config I provided. However, when replicating documents with attachments we do notice the following errors:
```
2017/06/16 08:52:07 Replicate: [db-local] stateFnActivePushBulkDocs got event: {PUSH_BULK_DOCS_SUCCEEDED [{com.id.dependency:pikaday_132  bad_request Missing data of attachment "pikaday.css"} {com.id.dependency:autocomplete_1  bad_request Missing data of attachment "autocomplete.css"} {com.id.dependency:moment_2103  bad_request Missing data of attachment "moment.min.js"} {com.id.dependency:fastclick_1  bad_request Missing data of attachment "fastclick.js"} {com.id.dependency:fontawesome_430  bad_request Missing data of attachment "fontawesome-webfont.eot"} {com.id.dependency:d3_1  bad_request Missing data of attachment "d3.js"} {com.id.dependency:pen-signature_1  bad_request Missing data of attachment "pen-signature.js"} {com.id.dependency:d3_2  bad_request Missing data of attachment "d3.js"} {com.id.dependency:charts_1  bad_request Missing data of attachment "charts.js"}]}
2017/06/16 08:52:07 Replicate: [db-local] BulkDoc response contains partial errors: [{Id:com.id.dependency:pikaday_132 Revision: Error:bad_request Reason:Missing data of attachment "pikaday.css"} {Id:com.id.dependency:autocomplete_1 Revision: Error:bad_request Reason:Missing data of attachment "autocomplete.css"} {Id:com.id.dependency:moment_2103 Revision: Error:bad_request Reason:Missing data of attachment "moment.min.js"} {Id:com.id.dependency:fastclick_1 Revision: Error:bad_request Reason:Missing data of attachment "fastclick.js"} {Id:com.id.dependency:fontawesome_430 Revision: Error:bad_request Reason:Missing data of attachment "fontawesome-webfont.eot"} {Id:com.id.dependency:d3_1 Revision: Error:bad_request Reason:Missing data of attachment "d3.js"} {Id:com.id.dependency:pen-signature_1 Revision: Error:bad_request Reason:Missing data of attachment "pen-signature.js"} {Id:com.id.dependency:d3_2 Revision: Error:bad_request Reason:Missing data of attachment "d3.js"} {Id:com.id.dependency:charts_1 Revision: Error:bad_request Reason:Missing data of attachment "charts.js"}], going to retry
```
This results in an infinite retry loop.

I ran a tcpdump capture on this request and when inspecting I noticed that the bulk put indeed only contains the json documents but no attachments in any multipart segment. Could this be a bug in sg-replicate?

The captured POST request can be found at https://gist.github.com/b-smets/3a46469c8a44673389670ec9651bbb36 .

**Comment by tleyden** [comment]
Re-opening this

**Comment by tleyden** [comment]
Moving to https://github.com/couchbase/sync_gateway/issues/2658

**Comment by ArihantRk** [comment]
@tleyden 
We tested http://latestbuilds.hq.couchbase.com/couchbase-sync-gateway/1.5.0/1.5.0-407/ patch and not able to reproduce Out of Memory issue in our 72 hrs traffic run. 
Below are the Observation is during traffic run
1. Document conflicts (revision limit is not working and older revisions (more than revision limit) are getting added) are happening due to SG Replication lagging and conflicts intern causes document size increase but did not observe out of memory issue in this patch. 
2. Sync Gateway memory increased from 2 GB to  5 GB over 72 hrs and we suspect it might be due to document size grow. 
Is this expected behavior? whether memory will keeps on increase if so what is the upper limit?

**Comment by tleyden** [comment]
> Document conflicts (revision limit is not working and older revisions (more than revision limit) are getting added) are happening due to SG Replication lagging and conflicts intern causes document size increase

Can you file a new ticket for that?  Definitely mention your topology (#sg and # of sg accel) and include your sync gateway config file, and steps to repro.

**Comment by tleyden** [comment]
> Is this expected behavior? whether memory will keeps on increase if so what is the upper limit?

If you are running in "channel cache" mode (non-sg accel), and if you have the revision cache set to a large number and have a lot of large docs, it would be expected to use a lot of memory.  You could try turning down the revision cache, but it's probably better to figure out how to avoid the conflicts and resulting large documents.

There's no upper limit, but a ticket was filed recently for that:

https://github.com/couchbase/sync_gateway/issues/2642

**Comment by nbrys** [comment]
Hi, we have retried this sync with the latest master build of sync gateway. The sync starts fine but quits after a couple of minutes with a clear error message.

Command:
curl -X POST --header 'Content-Type: application/json' --header 'Accept: application/json' -d '{ "source": "bucketa",  "target": "http://localhost:4985/bucketb" }' 'http://localhost:4985/_replicate'

We get the following message in the log:
{"error":"Internal Server Error","reason":"Internal error: Replication Aborted"}

The sync gateway log:
https://pastebin.com/AwnLMZVJ

**Comment by nbrys** [comment]
@tleyden ?

**Comment by tleyden** [comment]
@nbrys can you open a new ticket with:

* Clear steps to reproduce from scratch
* Actual behavior vs expected behavior
* Sync Gateway error logs, in particular logs that contain any error messages
* Kernel logs (if applicable), for example messages about being killed due to OOM
* Memory usage before being killed (if applicable, or possible)
